### PR TITLE
singleRowToOData(): fix $skiptoken handling

### PR DESCRIPTION
**Pull request by alxndrsn**
Correct spelling of `findIndex()`.

### CODE REVIEW

**Comment by alxndrsn** [comment]
@matthew-white @sadiqkhoja updated to include tests which fail without the change to `lib/formats/odata.js`:

```
  1) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 0,{"repeatId":"cf9a1b5cc83c6d6270c1eb98860d294eac5d526d"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  2) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 1,{"repeatId":"cf9a1b5cc83c6d6270c1eb98860d294eac5d526d"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  3) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 2,{"repeatId":"cf9a1b5cc83c6d6270c1eb98860d294eac5d526d"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  4) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for ,{"repeatId":"cf9a1b5cc83c6d6270c1eb98860d294eac5d526d"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  5) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 0,{"repeatId":"c76d0ccc6d5da236be7b93b985a80413d2e3e172"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  6) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 1,{"repeatId":"c76d0ccc6d5da236be7b93b985a80413d2e3e172"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  7) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for 2,{"repeatId":"c76d0ccc6d5da236be7b93b985a80413d2e3e172"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25

  8) odata message composition
       single submission output
         metadata generation
           with $skiptoken
             should return expected result for ,{"repeatId":"c76d0ccc6d5da236be7b93b985a80413d2e3e172"}:
     TypeError: filtered.fiindIndex is not a function
      at lib/formats/odata.js:513:25
```

File: test/unit/formats/odata.js [hunk 1]
```diff
@@ -930,6 +930,92 @@ describe('odata message composition', () => {
...
+            $top: 0,
+            skiptoken: { repeatId: blain.__id },
+            expectedNext: false,
+            expectedValue: [],
+          },
+          {
+            $top: 1,
+            skiptoken: { repeatId: blain.__id },
+            expectedNext: false,
+            expectedValue: [ blain ],
```
Comment (alxndrsn) [inline test/unit/formats/odata.js:973]: It looks like this result is inconsistent with a similar request for `repeatId: billy` above - here there is no `expectedNext`, whereas above the nextLink returned is a circular link to the original query.

File: test/unit/formats/odata.js [hunk 2]
```diff
@@ -930,6 +930,92 @@ describe('odata message composition', () => {
...
+          {
+            $top: 0,
+            skiptoken: { repeatId: billy.__id },
+            expectedNext: false,
+            expectedValue: [],
+          },
+          {
+            $top: 1,
+            skiptoken: { repeatId: billy.__id },
+            expectedNext: billy.__id,
```
Comment (alxndrsn) [inline test/unit/formats/odata.js:947]: It seems surprising that the nextLink actually points to the (last) value just returned.  Is this expected?

File: test/unit/formats/odata.js (hunk 1 above)
Comment (sadiqkhoja) [inline test/unit/formats/odata.js:973]:
```suggestion
            expectedValue: [ ],
```

File: test/unit/formats/odata.js [hunk 3]
```diff
@@ -930,6 +930,92 @@ describe('odata message composition', () => {
...
+            $top: 1,
+            skiptoken: { repeatId: blain.__id },
+            expectedNext: false,
+            expectedValue: [ blain ],
+          },
+          {
+            $top: 2,
+            skiptoken: { repeatId: blain.__id },
+            expectedNext: false,
+            expectedValue: [ blain ],
```
Comment (sadiqkhoja) [inline test/unit/formats/odata.js:979]:
```suggestion
            expectedValue: [ ],
```

File: test/unit/formats/odata.js [hunk 4]
```diff
@@ -930,6 +930,92 @@ describe('odata message composition', () => {
...
+          },
+          {
+            $top: undefined,
+            skiptoken: { repeatId: blain.__id },
+            expectedNext: false,
+            expectedValue: [ blain ],
```
Comment (sadiqkhoja) [inline test/unit/formats/odata.js:985]:
```suggestion
            expectedValue: [ ],
```

File: lib/formats/odata.js [hunk 5]
```diff
@@ -510,7 +510,7 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
     let { offset } = paging;
 
     if (skipToken) {
-      offset = filtered.fiindIndex(s => skipToken.repeatId === s.__id);
+      offset = filtered.findIndex(s => skipToken.repeatId === s.__id);
```
Comment (sadiqkhoja) [inline lib/formats/odata.js:513]: This is the source of the confusion and the bug, we need to send repeats after the one in `skipToken`.

```suggestion
      offset = filtered.findIndex(s => skipToken.repeatId === s.__id) + 1;
```

**Review by sadiqkhoja — COMMENTED** [review]
I think we should update the integration test (`test/integration/api/odata.js:312`) with the following as well.

```
 it('should limit and offset subtable results', testService(async (service) => {
      const asAlice = await withSubmission(service, identity);
      const nextlink = await asAlice.get("/v1/projects/1/forms/doubleRepeat.svc/Submissions('double')/children/child?$top=1&$skip=1")
        .expect(200)
        .then(({ body }) => {
          body.should.eql({
            '@odata.context': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/$metadata#Submissions.children.child',
            '@odata.nextLink': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/Submissions(%27double%27)/children/child?%24top=1&%24skiptoken=01eyJyZXBlYXRJZCI6ImI2ZTkzYTgxYTUzZWVkMDU2NmU2NWU0NzJkNGE0YjlhZTM4M2VlNmQifQ%3D%3D',
            value: [{
              __id: 'b6e93a81a53eed0566e65e472d4a4b9ae383ee6d',
              '__Submissions-id': 'double',
              name: 'Bob',
              toys: {
                'toy@odata.navigationLink': "Submissions('double')/children/child('b6e93a81a53eed0566e65e472d4a4b9ae383ee6d')/toys/toy"
              }
            }]
          });

          return body['@odata.nextLink'];
        });

      await asAlice.get(nextlink.replace('http://localhost:8989', ''))
        .expect(200)
        .then(({ body }) => {
          body.should.eql({
            '@odata.context': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/$metadata#Submissions.children.child',
            value: [{
              __id: '8954b393f82c1833abb19be08a3d6cb382171f54',
              '__Submissions-id': 'double',
              name: 'Chelsea',
              toys: {
                'toy@odata.navigationLink': "Submissions('double')/children/child('8954b393f82c1833abb19be08a3d6cb382171f54')/toys/toy"
              }
            }]
          });
        });
    }));
```

**Comment by alxndrsn** [comment]
@sadiqkhoja thanks for the corrections :pray:

**Comment by alxndrsn** [comment]
> I think we should update the integration test (`test/integration/api/odata.js:312`) with the following as well.
> 
> ```
>  it('should limit and offset subtable results', testService(async (service) => {
>       const asAlice = await withSubmission(service, identity);
>       const nextlink = await asAlice.get("/v1/projects/1/forms/doubleRepeat.svc/Submissions('double')/children/child?$top=1&$skip=1")
>         .expect(200)
>         .then(({ body }) => {
>           body.should.eql({
>             '@odata.context': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/$metadata#Submissions.children.child',
>             '@odata.nextLink': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/Submissions(%27double%27)/children/child?%24top=1&%24skiptoken=01eyJyZXBlYXRJZCI6ImI2ZTkzYTgxYTUzZWVkMDU2NmU2NWU0NzJkNGE0YjlhZTM4M2VlNmQifQ%3D%3D',
>             value: [{
>               __id: 'b6e93a81a53eed0566e65e472d4a4b9ae383ee6d',
>               '__Submissions-id': 'double',
>               name: 'Bob',
>               toys: {
>                 'toy@odata.navigationLink': "Submissions('double')/children/child('b6e93a81a53eed0566e65e472d4a4b9ae383ee6d')/toys/toy"
>               }
>             }]
>           });
> 
>           return body['@odata.nextLink'];
>         });
> 
>       await asAlice.get(nextlink.replace('http://localhost:8989', ''))
>         .expect(200)
>         .then(({ body }) => {
>           body.should.eql({
>             '@odata.context': 'http://localhost:8989/v1/projects/1/forms/doubleRepeat.svc/$metadata#Submissions.children.child',
>             value: [{
>               __id: '8954b393f82c1833abb19be08a3d6cb382171f54',
>               '__Submissions-id': 'double',
>               name: 'Chelsea',
>               toys: {
>                 'toy@odata.navigationLink': "Submissions('double')/children/child('8954b393f82c1833abb19be08a3d6cb382171f54')/toys/toy"
>               }
>             }]
>           });
>         });
>     }));
> ```

Is this a new test?  Could it be added as a separate PR?

**Comment by sadiqkhoja** [comment]
This is an existing test under `describe('api: /forms/:id.svc) -> describe('/Submissions(id)/… GET')` / line#312, but we can certainly have a separate PR for this.

**Review by sadiqkhoja — APPROVED** [review]

**Pull request merged**